### PR TITLE
install: cache the presets before evaluating

### PR DESCRIPTION
**Pull request by crawford**
The previous implementation traversed the various config directories,
walking the preset files and parsing each line to determine if a service
should be enabled or disabled. It did this for every service which
resulted in a combinatorial explosion.

This approach parses each of the preset entries into a hashmap which is
then used to check if each service should be enabled or disabled.

### CODE REVIEW

File: src/shared/install.c [hunk 1]
```diff
@@ -2367,17 +2370,19 @@ int unit_file_exists(UnitFileScope scope, const LookupPaths *paths, const char *
         return 1;
 }
 
-int unit_file_query_preset(UnitFileScope scope, const char *root_dir, const char *name) {
+static int unit_file_list_presets(UnitFileScope scope, const char *root_dir, OrderedHashmap **presets) {
         _cleanup_strv_free_ char **files = NULL;
+        OrderedHashmap *h;
```
Comment (vcaputo) [inline src/shared/install.c:2375]: _cleanup_...  so the error return paths free h, start as NULL...

File: src/shared/install.c [hunk 2]
```diff
@@ -2425,31 +2434,70 @@ int unit_file_query_preset(UnitFileScope scope, const char *root_dir, const char
...
 
-                        log_debug("Couldn't parse line '%s'", l);
+                        log_syntax(NULL, LOG_WARNING, *p, n, 0, "Couldn't parse line '%s'. Ignoring.", line);
                 }
         }
 
-        /* Default is "enable" */
-        log_debug("Preset file doesn't say anything about %s, enabling.", name);
-        return 1;
+        *presets = h;
```
Comment (vcaputo) [inline src/shared/install.c:2463]: assign NULL to h so the cleanup doesn't happen.

**Comment by keszybz** [comment]
Before we get into the specifics of implementation, let's talk about the design.

> if the service isn't found by hashmap_get, it falls back to iterating with fnmatch.

This does not sound right, because the order matters:

```
enable ab
disable a*
enable ac
```

The result here is different for ab and ac, even though they would both be found in a hashmap.

**Comment by poettering** [comment]
Yeah, there's globbing applied as well as ordering, and both make a hashmap unworkable I fear...

A "combinatorial explosion" sounds a bit strong wording to me... After all, it's just O(n*m) for a full system preset operation with n being  the number of unit files, and m the number of preset lines. Under the assumption that the latter is a constant, (of maybe 100 or less or so), this shouldn't be too bad, as it boils down to O(n) then, which is as good as it can get really...

How do your preset files look like? How large are they?

**Comment by crawford** [comment]
@keszybz yeah, your'e right. I'll add some tests to catch that failure. As @poettering said, I don't think the OrdererHashmap is the appropriate structure here. An in-order, linked-list representation of the presets should do the trick (like you originally suggested -_-).

> A "combinatorial explosion" sounds a bit strong wording to me...

@poettering maybe a bit of an exaggeration ;)  Admittedly, we have very few presets on CoreOS (by default, there is a single `disable *`), but I figured the more complex distros could benefit a bit from this speed up. The reason for this patch is to lay the foundation for another patch that I've been working on which allows instances, not just templates, to be specified in a preset file. I wanted to submit these independently since the optimization can stand on its own.

I'll take another stab at this over the weekend and rework the approach.

**Comment by keszybz** [comment]
Fedora currently has ~100 lines of presets, and atm 1553 unit files. So with an operation like preset-all we're looking at something that _could_ be optimized, as long as this doesn't introduce too much complexity.

>  An in-order, linked-list representation of the 

Please simply use an array. We have GREEDY_REALLOC and GREEDY_REALLOC0 which makes working with arrays a snap.

**Pull request closed without merging**